### PR TITLE
Baf 921/refactor usage of buffer

### PR DESCRIPTION
**Pull request by koudis**
Depends on:
https://github.com/bringauto/example-module/pull/5

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `Buffer` structure for enhanced buffer management.
  - Added new methods for buffer management, facilitating data transfer from device objects.

- **Enhancements**
  - Updated various methods across multiple classes to utilize the new `Buffer` type, improving type safety and clarity.
  - Improved initialization of member variables across several classes for better clarity.
  - Increased parallelism in the `main` function by introducing multiple context threads.

- **Bug Fixes**
  - Corrected data handling issues in `ErrorAggregator` and `ExternalConnection`.

- **Refactor**
  - Improved memory management by replacing the custom `struct buffer` with `Buffer`.
  - Simplified logging sink declarations in `initLogger` for enhanced clarity.

### CODE REVIEW

**Comment by koudis** [comment]
@coderabbitai review

File: source/bringauto/common_utils/ProtobufUtils.cpp [hunk 1]
```diff
@@ -97,4 +96,12 @@ ExternalProtocol::ExternalClient ProtobufUtils::createExternalClientCommandRespo
 	return externalMessage;
 }
 
-}
\ No newline at end of file
+void ProtobufUtils::copyStatusToBuffer(const InternalProtocol::DeviceStatus &status, bringauto::modules::Buffer &buffer) {
+	std::memcpy(buffer.getStructBuffer().data, status.statusdata().c_str(), status.statusdata().size());
```
Comment (koudis) [inline source/bringauto/common_utils/ProtobufUtils.cpp:100]: What if buffer is empty? What if data is nullptr?

Comment (koudis) [reply source/bringauto/common_utils/ProtobufUtils.cpp:100]: + check the statusdata.size() is lett than equal to struct buffer. Same fo function below pls.

File: source/bringauto/common_utils/ProtobufUtils.cpp [hunk 2]
```diff
@@ -73,15 +72,15 @@ ExternalProtocol::ExternalClient ProtobufUtils::createExternalClientStatus(const
...
+																		   const bringauto::modules::Buffer &errorMessage) {
 	ExternalProtocol::ExternalClient externalMessage;
 	ExternalProtocol::Status *status = externalMessage.mutable_status();
 	status->mutable_devicestatus()->CopyFrom(deviceStatus);
 	status->set_sessionid(sessionId);
 	status->set_devicestate(deviceState);
 	status->set_messagecounter(messageCounter);
-	if(errorMessage.size_in_bytes > 0 && errorMessage.data != nullptr) {
-		status->set_errormessage(errorMessage.data, errorMessage.size_in_bytes);
+	if(errorMessage.getStructBuffer().size_in_bytes > 0 && errorMessage.getStructBuffer().data != nullptr) {
```
Comment (koudis) [inline source/bringauto/common_utils/ProtobufUtils.cpp:82]: Maybe save erroBessage strus ::buffer into const variable? for better readability....

File: source/bringauto/external_client/connection/ExternalConnection.cpp [hunk 3]
```diff
@@ -432,26 +420,16 @@ std::vector<structures::DeviceIdentification> ExternalConnection::forceAggregati
 std::vector<structures::DeviceIdentification> ExternalConnection::getAllConnectedDevices() {
 	std::vector<structures::DeviceIdentification> devices {};
 	for(const auto &moduleNumber: settings_.modules) {
-		struct buffer unique_devices {};
-		int ret = moduleLibrary_.statusAggregators.at(moduleNumber)->get_unique_devices(&unique_devices);
+		std::list<structures::DeviceIdentification> unique_devices;
```
Comment (koudis) [inline source/bringauto/external_client/connection/ExternalConnection.cpp:423]: add {} pls.

File: source/bringauto/modules/ModuleHandler.cpp [hunk 4]
```diff
@@ -46,32 +45,25 @@ void ModuleHandler::handleMessages() {
...
 	}
 }
 
 void ModuleHandler::checkTimeoutedMessages(){
 	for (const auto& [key, statusAggregator] : moduleLibrary_.statusAggregators) {
+		auto moduleLibraryHandler = moduleLibrary_.moduleLibraryHandlers.at(key);
 		if(statusAggregator->getTimeoutedMessageReady()){
-			struct buffer unique_devices {};
-			int ret = statusAggregator->get_unique_devices(&unique_devices);
+			std::list<structures::DeviceIdentification> unique_devices;
```
Comment (koudis) [inline source/bringauto/modules/ModuleHandler.cpp:56]: add {}.
Add [] to all places which initialize C++/C structures pls.

File: source/bringauto/modules/ModuleHandler.cpp [hunk 5]
```diff
@@ -131,14 +118,13 @@ void ModuleHandler::handleDisconnect(const structures::DeviceIdentification& dev
 void ModuleHandler::sendAggregatedStatus(const structures::DeviceIdentification &deviceId, const ip::Device &device,
 										 bool disconnected) {
 	auto &statusAggregator = moduleLibrary_.statusAggregators.at(deviceId.getModule());
```
Comment (koudis) [inline source/bringauto/modules/ModuleHandler.cpp:120]: How do we know the given status aggregator is in the map struct?

File: source/bringauto/modules/ModuleManagerLibraryHandler.cpp [hunk 6]
```diff
@@ -113,4 +120,17 @@ void ModuleManagerLibraryHandler::deallocate(struct buffer *buffer){
 	deallocate_(buffer);
 }
 
-}
\ No newline at end of file
+bringauto::modules::Buffer ModuleManagerLibraryHandler::constructBuffer(std::size_t size) {
+	struct ::buffer buff;
+	buff.size_in_bytes = size;
+	if(allocate(&buff, size) != OK) {
+		throw std::bad_alloc();
```
Comment (koudis) [inline source/bringauto/modules/ModuleManagerLibraryHandler.cpp:127]: pls std::bad_alloc {} (I know the () are used in many places but {} implies the struct....)

File: source/bringauto/modules/ModuleManagerLibraryHandler.cpp [hunk 7]
```diff
@@ -113,4 +120,17 @@ void ModuleManagerLibraryHandler::deallocate(struct buffer *buffer){
...
+	}
+	return { buff, deallocate_ };
+}
+
+bringauto::modules::Buffer ModuleManagerLibraryHandler::constructBufferByTakeOwnership(struct ::buffer &buffer) {
+	return { buffer, deallocate_ };
```
Comment (koudis) [inline source/bringauto/modules/ModuleManagerLibraryHandler.cpp:133]: I think about checking the buffer itself. Maybe the check if the buffer is not empty (data != nullptr) should be added? To improve consistency and  avoidsegfault by a module impl. bug?

File: include/bringauto/modules/Buffer.hpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	~Buffer() = default;
+
+	Buffer& operator=(const Buffer& buff) = default;
+
+	/**
+	 * @brief Get a raw pointer to the buffer.
+	 *
+	 * @return pointer to a buffer
+	 */
+	inline struct ::buffer getStructBuffer() const { return *buffer_; }
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/modules/Buffer.hpp]: Check if buffer  is allocated added

File: include/bringauto/settings/Constants.hpp [hunk 9]
```diff
@@ -60,6 +60,11 @@ constexpr uint8_t header { 4 };
  */
 constexpr size_t buffer_length { 1024 };
 
+/**
+ * @brief maximal amount of external commands that can be stored in queue
+ */
+constexpr int max_external_commands { 3 };
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/settings/Constants.hpp]: Changed to unsigned int

File: source/bringauto/modules/ModuleManagerLibraryHandler.cpp (hunk 6 above)
Comment (MarioIvancik) [reply source/bringauto/modules/ModuleManagerLibraryHandler.cpp]: Changed braces

File: source/bringauto/modules/ModuleManagerLibraryHandler.cpp (hunk 7 above)
Comment (MarioIvancik) [reply source/bringauto/modules/ModuleManagerLibraryHandler.cpp:136]: Added check if buffer is allocated

File: include/bringauto/modules/ModuleManagerLibraryHandler.hpp [hunk 10]
```diff
@@ -31,31 +34,35 @@ class ModuleManagerLibraryHandler {
...
+	int generateFirstCommand(bringauto::modules::Buffer &default_command, unsigned int device_type);
 
-	int statusDataValid(const struct buffer status, unsigned int device_type);
+	int statusDataValid(const bringauto::modules::Buffer status, unsigned int device_type);
 
-	int commandDataValid(const struct buffer command, unsigned int device_type);
+	int commandDataValid(const bringauto::modules::Buffer command, unsigned int device_type);
 
-	int allocate(struct buffer *buffer_pointer, size_t size_in_bytes);
+	bringauto::modules::Buffer constructBuffer(std::size_t size = 0);
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/modules/ModuleManagerLibraryHandler.hpp]: Namespaces fixed

File: include/bringauto/modules/StatusAggregator.hpp [hunk 11]
```diff
@@ -73,21 +72,21 @@ class StatusAggregator {
 	 *
 	 * @see fleet-protocol/lib/module_gateway/include/status_aggregator.h
 	 */
-	int add_status_to_aggregator(const struct ::buffer status, const structures::DeviceIdentification& device);
+	int add_status_to_aggregator(const bringauto::modules::Buffer& status, const structures::DeviceIdentification& device);
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/modules/StatusAggregator.hpp]: namespaces fixed

File: source/bringauto/common_utils/ProtobufUtils.cpp [hunk 12]
```diff
@@ -97,4 +97,12 @@ ExternalProtocol::ExternalClient ProtobufUtils::createExternalClientCommandRespo
...
 }
 
-}
\ No newline at end of file
+void ProtobufUtils::copyStatusToBuffer(const InternalProtocol::DeviceStatus &status, bringauto::modules::Buffer &buffer) {
+	std::memcpy(buffer.getStructBuffer().data, status.statusdata().c_str(), status.statusdata().size());
+}
+
+void ProtobufUtils::copyCommandToBuffer(const InternalProtocol::DeviceCommand &command, bringauto::modules::Buffer &buffer) {
+	std::memcpy(buffer.getStructBuffer().data, command.commanddata().c_str(), command.commanddata().size());
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply source/bringauto/common_utils/ProtobufUtils.cpp:116]: added check if buffer is initialized

File: source/bringauto/common_utils/ProtobufUtils.cpp (hunk 1 above)
Comment (MarioIvancik) [reply source/bringauto/common_utils/ProtobufUtils.cpp:106]: added check if buffer is initialized

File: source/bringauto/common_utils/ProtobufUtils.cpp (hunk 2 above)
Comment (MarioIvancik) [reply source/bringauto/common_utils/ProtobufUtils.cpp]: added a new function that replaces this condition

File: source/bringauto/external_client/connection/ExternalConnection.cpp (hunk 3 above)
Comment (MarioIvancik) [reply source/bringauto/external_client/connection/ExternalConnection.cpp]: {} added

File: source/bringauto/modules/ModuleHandler.cpp (hunk 4 above)
Comment (MarioIvancik) [reply source/bringauto/modules/ModuleHandler.cpp]: {} added everywhere

File: source/bringauto/modules/ModuleHandler.cpp (hunk 5 above)
Comment (MarioIvancik) [reply source/bringauto/modules/ModuleHandler.cpp:120]: This function is being called only in handleDisconnect and handleStatus.
Both of those functions already check if that aggregator exists.

File: include/bringauto/settings/Constants.hpp [hunk 13]
```diff
@@ -60,6 +60,11 @@ constexpr uint8_t header { 4 };
  */
 constexpr size_t buffer_length { 1024 };
 
+/**
+ * @brief maximal amount of external commands that can be stored in queue
+ */
+constexpr unsigned int max_external_commands { 7 };
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/settings/Constants.hpp]: value changed back to 3

File: include/bringauto/modules/StatusAggregator.hpp [hunk 14]
```diff
@@ -73,21 +73,21 @@ class StatusAggregator {
...
-	int get_aggregated_status(struct ::buffer *generated_status, const structures::DeviceIdentification& device);
+	int get_aggregated_status(Buffer &generated_status, const structures::DeviceIdentification& device);
 
 	/**
 	 * @short Get all devices registered to aggregator
 	 *
 	 * @see fleet-protocol/lib/module_gateway/include/status_aggregator.h
 	 */
-	int get_unique_devices(struct ::buffer *unique_devices_buffer);
+	int get_unique_devices(std::list<structures::DeviceIdentification> &unique_devices_list);
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/modules/StatusAggregator.hpp:93]: documentation added

File: test/README.md [hunk 15]
```diff
@@ -1,7 +1,7 @@
 # Module Gateway - TESTS
 
 !!! NOTICE !!!
-Tests must be compiled with different compiler then **gcc**
+Tests must be compiled with different compiler than **gcc**
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply test/README.md]: warning changed specifically for gcc11

File: test/source/ErrorAggregatorTests.cpp [hunk 16]
```diff
@@ -1,82 +1,85 @@
...
+	std::memcpy(buffer.getStructBuffer().data, BUTTON_UNPRESSED, size);
 	return buffer;
 }
 
 void ErrorAggregatorTests::SetUp(){
-    auto libHandler = std::make_shared<bringauto::modules::ModuleManagerLibraryHandler>();
-	libHandler->loadLibrary(PATH_TO_MODULE);
-    errorAggregator.init_error_aggregator(libHandler);
+    libHandler_ = std::make_shared<bam::ModuleManagerLibraryHandler>();
+	libHandler_->loadLibrary(PATH_TO_MODULE);
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply test/source/ErrorAggregatorTests.cpp:19]: indentation fixed everywhere

File: source/bringauto/external_client/connection/communication/MqttCommunication.cpp [hunk 17]
```diff
@@ -113,7 +113,10 @@ std::shared_ptr<ExternalProtocol::ExternalServer> MqttCommunication::receiveMess
 	}
 
 	auto ptr = std::make_shared<ExternalProtocol::ExternalServer>();
-	ptr->ParseFromString(msg->get_payload_str());
+	if (!ptr->ParseFromArray(msg->get_payload_str().c_str(), static_cast<int>(msg->get_payload().size()))) {
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply source/bringauto/external_client/connection/communication/MqttCommunication.cpp]: made both get_payload_str

File: source/bringauto/external_client/ExternalClient.cpp [hunk 18]
```diff
@@ -144,12 +136,13 @@ bool ExternalClient::sendStatus(const structures::InternalClientMessage &interna
 			startExternalConnectSequence(connection);
 		}
 	} else {
+		auto errorBuffer = moduleLibrary_.moduleLibraryHandlers.at(moduleNumber)->constructBuffer();
 		bool ret = true;
 		if(internalMessage.disconnected()) {
-			connection.sendStatus(deviceStatus, ExternalProtocol::Status_DeviceState_DISCONNECT);
+			connection.sendStatus(deviceStatus, errorBuffer, ExternalProtocol::Status_DeviceState_DISCONNECT);
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply source/bringauto/external_client/ExternalClient.cpp]: errorBuffer is now optional in sendStatus

File: include/bringauto/structures/Connection.hpp [hunk 19]
```diff
@@ -45,16 +45,16 @@ struct Connection {
 		 * @brief message data
 		 */
 		std::vector<uint8_t> completeMessage {};
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/structures/Connection.hpp:47]: reserve removed, there was no point in using it

File: include/bringauto/modules/Buffer.hpp [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+#include <functional>
+
+
+
+namespace bringauto::modules {
+
+/**
+ * @brief Buffer structure used to simplify buffer management.
+ */
+struct Buffer final {
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/modules/Buffer.hpp:21]: description added

File: include/bringauto/modules/ModuleManagerLibraryHandler.hpp [hunk 21]
```diff
@@ -31,31 +34,35 @@ class ModuleManagerLibraryHandler {
...
+	int statusDataValid(const Buffer status, unsigned int device_type);
 
-	int commandDataValid(const struct buffer command, unsigned int device_type);
+	int commandDataValid(const Buffer command, unsigned int device_type);
 
-	int allocate(struct buffer *buffer_pointer, size_t size_in_bytes);
+	Buffer constructBuffer(std::size_t size = 0);
 
-	void deallocate(struct buffer *buffer);
+	Buffer constructBufferByTakeOwnership(struct ::buffer& buffer);
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/modules/ModuleManagerLibraryHandler.hpp]: documentation added

File: include/bringauto/external_client/connection/messages/SentMessagesHandler.hpp [hunk 22]
```diff
@@ -97,17 +97,18 @@ class SentMessagesHandler {
...
+	std::shared_ptr <structures::GlobalContext> context_ {};
 	/// Callback called by timer when status does not get response, registered by constructor
-	std::function<void()> endConnectionFunc_;
+	std::function<void()> endConnectionFunc_ {};
 
 	std::atomic<bool> responseHandled_ { false };
 
-	std::mutex responseHandledMutex_;
+	std::mutex responseHandledMutex_ {};
+	std::mutex ackMutex_ {};
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/external_client/connection/messages/SentMessagesHandler.hpp:112]: documentation added

File: source/bringauto/modules/ModuleHandler.cpp [hunk 23]
```diff
@@ -187,45 +173,43 @@ void ModuleHandler::handleStatus(const ip::DeviceStatus &status) {
...
-	}
-	std::memcpy(statusBuffer.data, statusData.c_str(), statusData.size());
+	auto statusBuffer = moduleHandler->constructBuffer(statusData.size());
+	common_utils::ProtobufUtils::copyStatusToBuffer(status, statusBuffer);
 
 	const auto deviceId = structures::DeviceIdentification(device);
 
-	struct ::buffer commandBuffer {};
-	int getCommandRc = statusAggregator->get_command(statusBuffer, deviceId, &commandBuffer);
+	if(statusBuffer.getStructBuffer().size_in_bytes == 0 ||
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply source/bringauto/modules/ModuleHandler.cpp]: isAllocated is now used

File: include/bringauto/modules/Buffer.hpp [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+		}
+		return raw_buffer_;
+	}
+
+	/**
+	 * @brief Determine if buffer is allocated.
+	 * 
+	 * @return true if buffer is allocated, false otherwise
+	 */
+	[[nodiscard]] bool isAllocated() const {
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply include/bringauto/modules/Buffer.hpp:52]: explanation added

File: source/bringauto/modules/StatusAggregator.cpp [hunk 25]
```diff
@@ -221,59 +190,34 @@ int StatusAggregator::update_command(const struct ::buffer command, const struct
...
-		return OK;
-	}
-
 	auto &deviceState = devices.at(device);
-	struct buffer generatedCommandBuffer {};
-	std::lock_guard<std::mutex> lock(mutex_);
-	auto &currCommand = devices.at(device).getCommand();
-	module_->generateCommand(&generatedCommandBuffer, status, deviceState.getStatus(), deviceState.getCommand(),
+	auto generatedCommandBuffer = module_->constructBuffer();
+	auto &currCommand = deviceState.getCommand();
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply source/bringauto/modules/StatusAggregator.cpp:209]: changed function name and documentation for better clarity as discussed

File: source/bringauto/structures/StatusAggregatorDeviceState.cpp [hunk 26]
```diff
@@ -7,48 +7,51 @@
...
 	command_ = commandBuffer;
 }
 
-const struct buffer &StatusAggregatorDeviceState::getCommand() const {
+const modules::Buffer &StatusAggregatorDeviceState::getCommand() {
+	if (!externalCommandQueue_.empty()) {
+		command_ = externalCommandQueue_.front();
+		externalCommandQueue_.pop();
+	}
 	return command_;
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply source/bringauto/structures/StatusAggregatorDeviceState.cpp]: changed function name and documentation for better clarity as discussed

File: source/bringauto/modules/StatusAggregator.cpp [hunk 27]
```diff
@@ -221,59 +190,34 @@ int StatusAggregator::update_command(const struct ::buffer command, const struct
 		return COMMAND_INVALID;
 	}
 
-	std::lock_guard<std::mutex> lock(mutex_);
-	devices.at(device).setCommand(command);
```
[the remark this answers is not in the document]
Comment (MarioIvancik) [reply source/bringauto/modules/StatusAggregator.cpp:225]: changed function name and documentation for better clarity as discussed

File: source/bringauto/external_client/ErrorAggregator.cpp [hunk 28]
```diff
@@ -22,85 +22,67 @@ int ErrorAggregator::destroy_error_aggregator() {
...
 
-	if(status.size_in_bytes == 0) {
+	if(status.getStructBuffer().size_in_bytes == 0) {
 		log::logWarning("Invalid status data for device: {}", device.convertToString());
 		return NOT_OK;
 	}
 
 	if(not devices_.contains(device)) {
 		devices_.insert({ device, {}});
+		devices_[device].lastStatus = module_->constructBuffer();
```
Comment (koudis) [inline source/bringauto/external_client/ErrorAggregator.cpp:38]: Are you sure we need this?

File: source/bringauto/external_client/connection/communication/MqttCommunication.cpp [hunk 29]
```diff
@@ -71,9 +71,9 @@ void MqttCommunication::initializeConnection() {
 
 void MqttCommunication::connect() {
 	client_ = std::make_unique<mqtt::async_client>(serverAddress_, clientId_,
-	                                               mqtt::create_options(MQTTVERSION_3_1_1, 20));
+												   mqtt::create_options(MQTTVERSION_3_1_1, 20));
```
Comment (koudis) [inline source/bringauto/external_client/connection/communication/MqttCommunication.cpp:74]: why tabulators?

File: include/bringauto/modules/ModuleManagerLibraryHandler.hpp [hunk 30]
```diff
@@ -31,33 +34,49 @@ class ModuleManagerLibraryHandler {
...
 
-	int generateCommand(struct buffer *generated_command, const struct buffer new_status,
-						const struct buffer current_status, const struct buffer current_command,
+	int generateCommand(Buffer &generated_command, const Buffer new_status,
+						const Buffer current_status, const Buffer current_command,
 						unsigned int device_type);
 
-	int aggregateStatus(struct buffer *aggregated_status, const struct buffer current_status,
-						const struct buffer new_status, unsigned int device_type);
+	int aggregateStatus(Buffer &aggregated_status, const Buffer current_status,
```
Comment (koudis) [inline include/bringauto/modules/ModuleManagerLibraryHandler.hpp:43]: Documentation is missing for these functions. for example: There are buffers but what happens when the can is unsuccessful etc.?

File: source/bringauto/external_client/ErrorAggregator.cpp (hunk 28 above)
Comment (MarioIvancik) [reply source/bringauto/external_client/ErrorAggregator.cpp]: code simplified

File: source/bringauto/external_client/connection/communication/MqttCommunication.cpp (hunk 29 above)
Comment (MarioIvancik) [reply source/bringauto/external_client/connection/communication/MqttCommunication.cpp]: indentation fixed

File: include/bringauto/modules/ModuleManagerLibraryHandler.hpp (hunk 30 above)
Comment (MarioIvancik) [reply include/bringauto/modules/ModuleManagerLibraryHandler.hpp:52]: documentation added

**Review by MarioIvancik — APPROVED** [review]

**Pull request merged**